From 71c0ea0e0717523f69f74a939976db1a4198d36c Mon Sep 17 00:00:00 2001
From: stasson <sebastien.tasson@intel.com>
Date: Wed, 20 Nov 2013 18:03:42 +0100
Subject: Fix Thread's cardImmuneLimit mismatch.

BZ: 152029

This is a fix for a regression in the "conditional card marking" patch.
(see Change 129488.)

At thread startup, thread->cardImmuneLimit must be refreshed just when
being added to the thread list while still protected by the thread list
mutex. If failing to do so, some objects may escape GC dirty marking
phase and be wrongly sweeped.

Change-Id: I97b99e12cd2d31ca9adb2b1be805ae93ffc51790
Category: device enablement
Domain: AOSP.Dalvik-Runtime
Origin: internal
Upstream-Candidate: no, need rework
Signed-off-by: stasson <sebastien.tasson@intel.com>
Signed-off-by: Ivan Popov <ivan.g.popov@intel.com>
---
 vm/Thread.cpp |   18 +++++++++++++++++-
 1 files changed, 17 insertions(+), 1 deletions(-)

diff --git a/vm/Thread.cpp b/vm/Thread.cpp
index 9fb553e..5e807c9 100644
--- a/vm/Thread.cpp
+++ b/vm/Thread.cpp
@@ -926,7 +926,7 @@ static bool prepareThread(Thread* thread)
 
 #ifdef WITH_CONDMARK
     /*
-     *  Refresh self->cardImmuneLimit;
+     *  Initlialize thread->cardImmuneLimit;
      */
     thread->cardImmuneLimit = gDvm.cardImmuneLimit;
 #endif
@@ -1511,6 +1511,14 @@ static void* interpThreadStart(void* arg)
     while (self->status != THREAD_VMWAIT)
         pthread_cond_wait(&gDvm.threadStartCond, &gDvm.threadListLock);
 
+#ifdef WITH_CONDMARK
+    /*
+     *  cardImmuneLimit may have change before thread was added to the list,
+     *  so refresh it here before releasing the lock.
+     */
+    self->cardImmuneLimit = gDvm.cardImmuneLimit;
+#endif
+
     dvmUnlockThreadList();
 
     /*
@@ -1843,6 +1851,14 @@ bool dvmAttachCurrentThread(const JavaVMAttachArgs* pArgs, bool isDaemon)
     if (!isDaemon)
         gDvm.nonDaemonThreadCount++;
 
+#ifdef WITH_CONDMARK
+    /*
+     *  cardImmuneLimit may have change before thread was added to the list,
+     *  so refresh it here before releasing the lock.
+     */
+    self->cardImmuneLimit = gDvm.cardImmuneLimit;
+#endif
+
     dvmUnlockThreadList();
 
     /*
-- 
1.7.4.1

